### PR TITLE
Integrate OAuth 0.13.0

### PR DESCRIPTION
**Pull request by mstruk**
### Type of change

- Enhancement / new feature
- Documentation

### Description

Integrate new features / config options from OAuth 0.13.0

- Keycloak authorization now uses a new `Authorizer` class that supports KRaft mode as well as Zookeeper mode.
- New config options were introduced to `keycloak` authorization:
  - `grantsMaxIdleSeconds` to control the removal of idle grants from the cache
  - `grantsGcPeriodSeconds` to control the periodicity of the background job that removes idle grant from the cache
  - `grantsAlwaysLatest` to control if latest grants should always be fetched from Keycloak at the start of a new session
- Environment variable `STRIMZI_OAUTH_METRIC_REPORTERS` can now be used similarly to `metric.reporters` to list the reporters to be instantiated and integrated with OAuth metrics. By default `JmxReporter` is automatically used. If `STRIMZI_OAUTH_METRIC_REPORTERS` is set, only the listed reporter classes will be used.
- An issue with processing JsonPath filter queries used in `oauth` authentication `customClaimCheck` configuration was fixed by updating the dependency library. Comparing nonexistent attribute against `null` using equals now evaluates differently. Check your configuration. Avoid using `null` in queries.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java [hunk 1]
```diff
@@ -138,6 +142,38 @@ public void setGrantsRefreshPoolSize(Integer grantsRefreshPoolSize) {
         this.grantsRefreshPoolSize = grantsRefreshPoolSize;
     }
 
+    @Description("The time in seconds after which an idle grant in the cache can be evicted. The fefault value is 300.")
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java:145]: Typo "fefault"

**Review by scholzj — COMMENTED** [review]
Is there some new feature notable enough to add it to the CHANGELOG.md?

**Comment by mstruk** [comment]
> Is there some new feature notable enough to add it to the CHANGELOG.md?

We might want to mention that users of OAuth authentication customClaimCheck option may want to make sure they don't use comparison with `null` in their JsonPath filter queries, because if used it was apparently relying on a buggy behaviour which is now fixed.

**Comment by scholzj** [comment]
> > Is there some new feature notable enough to add it to the CHANGELOG.md?
> 
> We might want to mention that users of OAuth authentication customClaimCheck option may want to make sure they don't use comparison with `null` in their JsonPath filter queries, because if used it was apparently relying on a buggy behaviour which is now fixed.

Right, that sounds like something you can add to the `Changes, deprecations and removals` section.

**Review by PaulRMellor — APPROVED** [review]
Looks god, @mstruk 
I left a couple of suggestions, including a comment on a description I wasn't entirely sure about

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java [hunk 2]
```diff
@@ -138,6 +142,38 @@ public void setGrantsRefreshPoolSize(Integer grantsRefreshPoolSize) {
...
+    @JsonProperty(defaultValue = "300")
+    public Integer getGrantsMaxIdleTimeSeconds() {
+        return grantsMaxIdleTimeSeconds;
+    }
+
+    public void setGrantsMaxIdleTimeSeconds(Integer grantsMaxIdleTimeSeconds) {
+        this.grantsMaxIdleTimeSeconds = grantsMaxIdleTimeSeconds;
+    }
+
+    @Description("A period in seconds in which cleaning of stale grants from cache is performed. The default value is 300.")
```
[the remark this answers is not in the document]
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java:156]: There is a job that runs every 300 seconds and performs cleaning. Thus, it runs periodically that's why I used a word period. Would another work be more fitting?

File: documentation/assemblies/metrics/assembly-metrics.adoc [hunk 3]
```diff
@@ -10,7 +10,7 @@ Collecting metrics is critical for understanding the health and performance of y
 By monitoring metrics, you can actively identify issues before they become critical and make informed decisions about resource allocation and capacity planning. Without metrics, you may be left with limited visibility into the behavior of your Kafka deployment, which can make troubleshooting more difficult and time-consuming. Setting up metrics can save you time and resources in the long run, and help ensure the reliability of your Kafka deployment.
 
 Metrics are available for each Strimzi component, allowing you to gain insights into their individual performance. 
-Additionally, you can collect metrics specific to `oauth` authentication and `opa` or `keycloak` authorization by enabling the `enableMetrics` property in the listener configuration of the `Kafka` resource. 
+Additionally, you can collect metrics specific to `oauth` authentication and `opa` or `keycloak` authorization by enabling the `enableMetrics` property in the listener or authorization configuration of the `Kafka` resource.
```
[the remark this answers is not in the document]
Comment (mstruk) [reply documentation/assemblies/metrics/assembly-metrics.adoc:13]: This is not actually new. It's an omission I noticed in the docs.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAuthorizationKeycloak.java]: @PaulRMellor Does the latest proposed wording sound right, or it needs some small modification still?

File: documentation/assemblies/metrics/assembly-metrics.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaulRMellor) [reply documentation/assemblies/metrics/assembly-metrics.adoc:13]: Yeah. Maybe this is for another PR.

File: CHANGELOG.md [hunk 4]
```diff
@@ -13,13 +13,15 @@
 * Update OpenTelemetry 1.19.0
 * Fixed ordering of JVM performance options [#8579](https://github.com/strimzi/strimzi-kafka-operator/issues/8579)
 * Log a warning when a KafkaTopic has no spec [#8465](https://github.com/strimzi/strimzi-kafka-operator/issues/8465)
+* Update Strimzi OAuth library to 0.13.0
```
Comment (ppatierno) [inline CHANGELOG.md:16]: Is there any special thing, bug fixed or new feature that we should mention here as sub-topics?

Comment (mstruk) [reply CHANGELOG.md:16]: Maybe we could say:
```suggestion
* Updated Strimzi OAuth library to 0.13.0 with better support for KRaft, and some new features and bug fixes.
```

Comment (mstruk) [reply CHANGELOG.md:16]: Actually I can improve the Description in this PR ... describe the added config options, and add a link to PR:
```suggestion
* Updated Strimzi OAuth library to 0.13.0 [#8802](https://github.com/strimzi/strimzi-kafka-operator/issues/8802)
```

Comment (scholzj) [reply CHANGELOG.md:16]: I think it is better to use the first one rather than link to a PR.

Comment (scholzj) [reply CHANGELOG.md:16]: That said, `some new features and bug fixes` is more or less meaning less. So I think you do not need to add it. Also, is the KRaft support via the operator actually functional? Did you tried it?

Comment (ppatierno) [reply CHANGELOG.md:16]: Agree, I would leave `some new features and bug fixes` out unless you list all of them :-)
I would leave just the the one related to KRaft.

Comment (mstruk) [reply CHANGELOG.md:16]: > Also, is the KRaft support via the operator actually functional? Did you tried it?
That's a good point. Let me try it.

Comment (mstruk) [reply CHANGELOG.md:16]: Works like a charm.

Comment (scholzj) [reply CHANGELOG.md]: Maybe you can in that case use something like this in the CHANGELOG?
```
* Updated Strimzi OAuth library to 0.13.0 with KRaft support
```

Comment (scholzj) [reply CHANGELOG.md]: Oh, I see ou already pushed it ... it looks fine. No need to use the suggestion above.

**Review by scholzj — APPROVED** [review]

**Review by ppatierno — APPROVED** [review]

**Pull request merged**